torch/_inductor/codecache.py

```diff
@@ -379,6 +386,240 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
     tmp_path.rename(path)
 
 
+@dataclasses.dataclass
+class TensorMetadata:
+    """
+    The Tensor metadata relevant when hashing FxGraph cache keys.
+    """
+
+    dtype: torch.dtype
+    shape: torch.Size
+    stride: Tuple[Any, ...]
+    device: torch.device
+    requires_grad: bool
+    storage_offset: int
+    memory_format: Optional[torch.memory_format]
+    is_quantized: bool
+
+
+@dataclasses.dataclass
+class TensorMetadataAndValues:
+    """
+    TensorMetadata plus the elements as a list of raw values.
+    """
+
+    tensor_metadata: TensorMetadata
+    values: List[Any]
+
+
+def extract_tensor_metadata(t: torch.Tensor) -> TensorMetadata:
+    """
+    Extract the TensorMetadata of a tensor.
+    """
+    # Note the order here matters since is_contiguous() can be True
```

Comment:
I figured out why I was occasionally seeing different memory formats on consecutive runs. It's possible for is_contiguous to match > 1 of the memory formats options. So the order here matters and this iterable should not be a set.